File: frontend/src/engine/theme.ts
// Ported from CausewaybayGolang/typescript/src/engine/theme.ts (which in turn
// carries it over from that project's love2d Lua). Kept close to the original
// so a fix made there can still be read across; Causewaybay Hacker changes are
// marked where they occur.
/**
 * Super Mario World sky and Wonder Boy candy, not neon cyberpunk.
 * The palette of `love2d/src/theme.lua`, as numbers a canvas can use.
 */
export type RGBA = readonly [number, number, number, number];

const c = (r: number, g: number, b: number, a = 1): RGBA => [r / 255, g / 255, b / 255, a];

export const Theme = {
  void: c(20, 28, 72),
  sky: c(92, 148, 252),
  navy: c(28, 36, 92),
  panel: c(248, 208, 136),
  wood: c(176, 104, 40),
  coin: c(248, 208, 48),
  brick: c(200, 76, 12),
  grass: c(0, 168, 0),
  red: c(216, 40, 0),
  cyan: c(80, 216, 248),
  pink: c(248, 120, 168),
  cream: c(252, 236, 200),
  ink: c(40, 24, 16),
  admit: c(0, 168, 68),
  dim: c(120, 104, 88),
  // Causewaybay Hacker change: the panel face was 0.94 opaque, which was right
  // when there was nothing behind it. With the city on the WebGL layer the
  // panels are glass over neon — enough to read text on, not so much that the
  // three.js work is invisible on every screen but the map.
  paper: [0.1, 0.09, 0.24, 0.84] as RGBA,

  landW: 1280,
  landH: 720,
  portW: 720,
  portH: 1280,
};

/** Each track's colour on the map buttons, and the haze over the overworld. */
export const TRACK_COL: Record<string, RGBA> = {
  go: Theme.cyan,
  rust: [0.95, 0.47, 0.16, 1],
  python: [0.36, 0.62, 0.92, 1],
};

export const TRACK_HAZE: Record<string, RGBA> = {
  go: [0.02, 0.02, 0.1, 0.22],
  rust: [0.32, 0.08, 0.02, 0.26],
  python: [0.02, 0.1, 0.22, 0.28],
};

/** A colour a canvas will take, with an optional alpha override. */
export function css(col: RGBA, alpha?: number): string {
  const a = alpha ?? col[3] ?? 1;
  const r = Math.round(col[0] * 255);
  const g = Math.round(col[1] * 255);
  const b = Math.round(col[2] * 255);
  return `rgba(${r},${g},${b},${a})`;
}

export function withAlpha(col: RGBA, a: number): RGBA {
  return [col[0], col[1], col[2], a];
}

File: love2d/src/theme.lua
-- The palette, as numbers.
--
-- Ported from `CausewaybayGolang/love2d/src/theme.lua`, which is the same
-- table `CausewaybayGolang/typescript/src/engine/theme.ts` carries, and which
-- `docs/art.md` §2 names as the register for this game too: "Super Mario
-- World sky and Wonder Boy candy, not neon cyberpunk."
--
-- Added here: the land tints from art.md §2, which are what separates
-- RUST LAND from GO LAND without a second set of sprites — and, since the
-- same trick scales, C++ LAND (ISO C++ blue, #00599C) and PYTHON LAND
-- (Python gold, #FFD43B) from both.

local T = {}

T.void = { 20 / 255, 28 / 255, 72 / 255, 1 }
T.sky = { 92 / 255, 148 / 255, 252 / 255, 1 }
T.navy = { 28 / 255, 36 / 255, 92 / 255, 1 }
T.panel = { 248 / 255, 208 / 255, 136 / 255, 1 }
T.wood = { 176 / 255, 104 / 255, 40 / 255, 1 }
T.coin = { 248 / 255, 208 / 255, 48 / 255, 1 }
T.brick = { 200 / 255, 76 / 255, 12 / 255, 1 }
T.grass = { 0 / 255, 168 / 255, 0 / 255, 1 }
T.red = { 216 / 255, 40 / 255, 0 / 255, 1 }
T.cyan = { 80 / 255, 216 / 255, 248 / 255, 1 }
T.pink = { 248 / 255, 120 / 255, 168 / 255, 1 }
T.cream = { 252 / 255, 236 / 255, 200 / 255, 1 }
T.ink = { 40 / 255, 24 / 255, 16 / 255, 1 }
T.admit = { 0 / 255, 168 / 255, 68 / 255, 1 }
T.dim = { 120 / 255, 104 / 255, 88 / 255, 1 }

-- art.md §2: the land tint is a haze over the map and a border on the panels.
T.land = {
  rust = { 0.95, 0.47, 0.16, 1 },
  go = { 80 / 255, 216 / 255, 248 / 255, 1 },
  cpp = { 0 / 255, 89 / 255, 156 / 255, 1 },
  python = { 255 / 255, 212 / 255, 59 / 255, 1 },
}
-- The haze is the tint at map strength: a deep blue over the typhoon
-- shelter at noon, a warm amber over the wet market at dawn. Python's is
-- the faintest because gold over a whole plate reads as a sepia filter
-- before it reads as a colour.
T.haze = {
  rust = { 0.95, 0.47, 0.16, 0.16 },
  go = { 80 / 255, 216 / 255, 248 / 255, 0.14 },
  cpp = { 0 / 255, 89 / 255, 156 / 255, 0.18 },
  python = { 255 / 255, 212 / 255, 59 / 255, 0.12 },
}

-- The editor's colours, keyed by `src/editor.lua`'s span kinds. Chosen out of
-- the palette above rather than from a syntax theme, so the editor looks like
-- the rest of the game and not like an IDE somebody embedded.
T.code = {
  text = T.cream,
  keyword = { 248 / 255, 120 / 255, 168 / 255, 1 },
  type = { 80 / 255, 216 / 255, 248 / 255, 1 },
  string = { 152 / 255, 216 / 255, 120 / 255, 1 },
  number = T.coin,
  comment = { 140 / 255, 128 / 255, 112 / 255, 1 },
  macro = { 0.95, 0.62, 0.30, 1 },
  punct = { 208 / 255, 196 / 255, 172 / 255, 1 },
}

-- The verdict colours from §5.4's closed set.
T.verdict = {
  accepted = T.admit,
  wrong_answer = T.brick,
  compile_error = T.red,
  runtime_error = T.red,
  timeout = T.coin,
  output_limit = T.coin,
  internal_error = T.dim,
}

T.landW, T.landH = 1280, 720
T.portW, T.portH = 720, 1280

function T.withAlpha(c, a)
  return { c[1], c[2], c[3], a }
end

return T
